app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -546,4 +461,167 @@ public int setSelectionExpandWholeLines() {
+            } else if (_maxNumber < 100) {
+                _maxNumberDigits = 2;
+            } else if (_maxNumber < 1000) {
+                _maxNumberDigits = 3;
+            } else if (_maxNumber < 10000) {
+                _maxNumberDigits = 4;
+            } else {
+                _maxNumberDigits = 5;
+            }
+            return _maxNumberDigits != oldDigits;
+        }
+
+        public boolean isOutOfLineNumbersArea() {
+            if (_visibleArea.top > _lineNumbersArea.top && _visibleArea.bottom < _lineNumbersArea.bottom) {
+                return false;
+            } else {
+                // If current visible area is out of line numbers area, line numbers area will be updated
+                _lineNumbersArea.top = _visibleArea.top - _visibleArea.height();
+                _lineNumbersArea.bottom = _visibleArea.bottom + _visibleArea.height();
+                return true;
+            }
+        }
+
+        public void startLineTracking() {
+            _editor.removeTextChangedListener(_lineTrackingWatcher);
+            _maxNumber = 1;
+            final CharSequence text = _editor.getText();
+            if (text != null) {
+                _maxNumber += TextViewUtils.countChar(text, 0, text.length(), '\n');
+            }
+            _editor.addTextChangedListener(_lineTrackingWatcher);
+        }
+
+        public void stopLineTracking() {
+            _editor.removeTextChangedListener(_lineTrackingWatcher);
+        }
+
+        /**
+         * Draw line numbers.
+         *
+         * @param canvas the canvas on which the line numbers will be drawn.
+         */
+        public void draw(final Canvas canvas) {
+            if (!_editor.getLocalVisibleRect(_visibleArea)) {
+                return;
+            }
+
+            final CharSequence text = _editor.getText();
+            final Layout layout = _editor.getLayout();
+            if (text == null || layout == null) {
+                return;
+            }
+
+            // Only if the text size or the max line number of digits changed,
+            // update the gutter's parameters and reset padding
+            if (isTextSizeChanged() || isMaxNumberDigitsChanged()) {
+                _numberX = LINE_NUMBER_PADDING_LEFT + Math.round(_paint.measureText(String.valueOf(_maxNumber)));
+                _gutterX = _numberX + LINE_NUMBER_PADDING_RIGHT;
+                _editor.setPadding(_gutterX + 10, _editor.getPaddingTop(), _editor.getPaddingRight(), _editor.getPaddingBottom());
+            }
+
+            // Draw the right border of the gutter
+            _paint.setColor(Color.LTGRAY);
+            canvas.drawLine(_gutterX, _lineNumbersArea.top, _gutterX, _lineNumbersArea.bottom, _paint);
+
+            int i = _startLine[0], number = _startLine[1];
+            // Only if the visible area is out of current line numbers area,
+            // update the parameters of the line numbers which will be drawn
+            if (isOutOfLineNumbersArea()) {
```

Comment:
This will be triggered only if there is a shift right? Not if highlighting state changes or we change text size etc etc (all of which can affect Y Positions).

Do line numbers draw correctly when you toggle highlighting in a todo.txt file with just 2-3 lines

Reply:
Yes, it behaved well in these situations. It does not save fixed y positions, so it can access and use new y positions of lines if the they adjusted.